Extension traits (#2812)
google/comprehensive-rust: src/idiomatic/leveraging-the-type-system/extension-traits/extending-other-traits.md

---
minutes: 15
---

> # Extending Other Traits

Reviewer: It might be better to call this slide "Blanket Impls" since that's what this slide is demonstrating. At least personally, I don't think of blanket impls as "extending other traits", though that does seem like a reasonable way to look at it. But blanket impls are also generally useful outside of extension traits, so I think it's worth talking about blanket impls generally beyond their use in extension traits.

Author: I see where you're coming from, but I disagree—blanket impls are a building block for this pattern, but they are a side character in this slide.
We could change the ToC to reserve some space for blanket impls, but then I would want to touch a variety of topics that we don't even mention here (e.g., overlap risk, common use cases to ease ergonomics, such as implementing traits for functions, etc.).

Reviewer 2: > I don't think of blanket impls as "extending other traits"

I think this is a matter of perspective and background. "Protocol extensions" in Swift are a big deal (ex: [1](https://medium.com/@leandromperez/protocol-extensions-gotcha-9ef1a42c83b6)) and they are presented as extensions on the protocol, like "extending other traits" here.

Reviewer: Totally fair! I'm fine to leave things as-is.

> Extension traits can attach new methods to _all_ implementors of a given trait:

Reviewer 2: Like on the other slide, I'd suggest to start by explaining the desired result first.

Author: Done in 771a37a8

```rust
mod ext {
    use std::fmt::Display;

    pub trait DisplayExt {
        fn quoted(&self) -> String;
    }

    impl<T: Display> DisplayExt for T {
        fn quoted(&self) -> String {
            format!("'{}'", self)
        }
    }
}

pub use ext::DisplayExt as _;

assert_eq!("dad".quoted(), "'dad'");
assert_eq!(4.quoted(), "'4'");

> assert_eq!(true.quoted(), "'true'");

Reviewer 2: This isn't runnable, is it?

Author: It is, why not?

Reviewer 2: Because the `assert_eq!` should be in a function.

Author: The Rust playground will implicitly wrap them in a `main` function!

  [_blanket implementations_][1].

  Blanket implementations allow us to implement a trait for a generic type `T`,
  as long as it satisfies the trait bounds specified in the `impl` block. In
  this case, the only requirement is that `T` implements the `Display` trait.

- Conventionally, the extension trait is named after the trait it extends,
  following by the `Ext` suffix. In the example above, `DisplayExt`.

- There are entire libraries aimed at extending foundational traits with new
  functionality.

  [`itertools`] provides a wide range of iterator adapters and utilities via the
  [`Itertools`] trait. [`futures`] provides [`FutureExt`] to extend the
  [`Future`] trait.

## More To Explore

- Extension traits can be used by libraries to distinguish between stable and
  experimental methods.

  Stable methods are part of the trait definition.

  Experimental methods are provided via an extension trait defined in a
  different library, with a less restrictive stability policy. Some utility
  methods are then "promoted" to the core trait definition once they have been
  proven useful and their design has been refined.

> - Extension traits can be used to split a [dyn-incompatible trait][2] in two:

Reviewer 2: This is an excellent use case, but without a code example it is, most likely, impossible to learn from a mere description. The only type of student who would be satisfied with a terse description is someone who is already familiar with the idea.

Author: I placed this under "More to explore" to signal that the instructor can decide whether to mention/explain/dive into this topic.
I can expand it with code examples and more details, but then it'd make more sense to extract it into its own slide.
Alternatively, we can keep the terse explanation and provide a link to a more in-depth resource for those who want to dig deeper.

---
minutes: 15
---

# Extending Other Traits

As with types, it may be desirable to **extend foreign traits**. In particular,
to attach new methods to _all_ implementors of a given trait.

```rust
mod ext {
    use std::fmt::Display;

    pub trait DisplayExt {
        fn quoted(&self) -> String;
    }

    impl<T: Display> DisplayExt for T {
        fn quoted(&self) -> String {
            format!("'{}'", self)
        }
    }
}

pub use ext::DisplayExt as _;

assert_eq!("dad".quoted(), "'dad'");
assert_eq!(4.quoted(), "'4'");
assert_eq!(true.quoted(), "'true'");
```

<details>

- Highlight how we added new behaviour to _multiple_ distinct types at once.
  `.quoted()` can be called on string slices, numbers and booleans since they
  all implement the `Display` trait.

  This flavour of the extension trait pattern is built on top of
  [_blanket implementations_][1].

  Blanket implementations allow us to implement a trait for a generic type `T`,
  as long as it satisfies the trait bounds specified in the `impl` block. In
  this case, the only requirement is that `T` implements the `Display` trait.

> - Draw the students attention to the implementation of `DisplayExt::quoted`: we

Reviewer 2: ```suggestion
- Draw the students' attention to the implementation of `DisplayExt::quoted`: we
```
possessive

>   can't make any assumptions about the type of `T` other than that it implements

Reviewer 2: ```suggestion
  can't make any assumptions about `T` other than that it implements
```

  `Display`. All our logic must either use methods from `Display` or

>   functions/macros that doesn't require `T` to implement any other trait.

Reviewer 2: ```suggestion
  functions/macros that don't require other traits..
```
"functions/macros" is plural + a concise rewrite.

> 

Reviewer 2: ```suggestion

  For example, we can call `format!` with `T`, but can't call `.to_uppercase()` because it is not necessarily a `String`.
  
```

  We could introduce additional trait bounds on `T`, but it would restrict the
  set of types that can leverage the extension trait.

- Conventionally, the extension trait is named after the trait it extends,

>   following by the `Ext` suffix. In the example above, `DisplayExt`.

Reviewer: I'm not sure I agree with this, I think personally I'd name the trait something like `Quote` or `Quoted`. Calling it `DisplayExt` is reasonable if this is the only trait extending `Display`, but what if you had multiple traits extending `Display` in different ways? For example, I might have another trait that has a blanket impl on `T: Display + Clone` or something. Calling that `DisplayCloneExt` would be cumbersome and doesn't describe what behavior it's extending those traits with.

>   following by the `Ext` suffix. In the example above, `DisplayExt`.

Reviewer 2: ```suggestion
  followed by the `Ext` suffix. In the example above, `DisplayExt`.
```

> - There are entire libraries aimed at extending foundational traits with new
>   functionality.
> 
>   [`itertools`] provides a wide range of iterator adapters and utilities via the
>   [`Itertools`] trait. [`futures`] provides [`FutureExt`] to extend the
>   [`Future`] trait.

Reviewer 2: ```suggestion
- There are entire crates that extend standard library traits with new functionality.

  - `itertools` crate provides the `Itertools` trait that extends `Iterator`. It adds many iterator adapters, such as `interleave` and `unique`. It provides new algorithmic building blocks for iterator pipelines built with method chaining.
  
  - `futures` crate provides the `FutureExt` trait, which extends the `Future` trait with new combinators and helper methods.
```

More structure with a nested list + a bit more content.

## More To Explore

- Extension traits can be used by libraries to distinguish between stable and
  experimental methods.

  Stable methods are part of the trait definition.

  Experimental methods are provided via an extension trait defined in a
  different library, with a less restrictive stability policy. Some utility
  methods are then "promoted" to the core trait definition once they have been
  proven useful and their design has been refined.

- Extension traits can be used to split a [dyn-incompatible trait][2] in two:

  - A **dyn-compatible core**, restricted to the methods that satisfy
    dyn-compatibility requirements.
  - An **extension trait**, containing the remaining methods that are not

>     dyn-compatible. (e.g., methods with a generic parameter).

Reviewer 2: ```suggestion
    dyn-compatible (e.g., methods with a generic parameter).
```
The parenthetical continues the sentence.
